apis/python/src/tiledbsoma/_query.py

```diff
@@ -465,9 +469,29 @@ def to_anndata(
 
         tp = self._threadpool
         x_collection = self._ms.X
-        all_x_names = [X_name, *list(X_layers)]
         all_x_arrays: dict[str, SparseNDArray] = {}
-        for _xname in all_x_names:
+        # Enforce identical semantics to io.to_anndata for X handling
+        if X_name is None and X_layers:
+            raise ValueError("If X_name is None, X_layers must not be provided")
+
+        chosen_x: str | None = None
+        if X_name is MISSING:
+            if "data" in x_collection:
+                chosen_x = "data"
+        elif X_name is not None:
+            if not isinstance(X_name, str) or not X_name:
+                raise ValueError("X layer names must be specified as a string.")
+            if X_name not in x_collection:
+                raise ValueError(f"X layer name '{X_name}' not found in measurement")
+            chosen_x = X_name
+
+        # Gather X and any requested extra layers
+        for _xname in [chosen_x] if chosen_x is not None else []:
+            x_array = x_collection[_xname]
+            if not isinstance(x_array, SparseNDArray):
+                raise NotImplementedError("Dense array unsupported")
+            all_x_arrays[_xname] = x_array
```

Comment:
This piece of code is hard to reason about. You don't need to create a list and an indirection with `_xname`, just use `chosen_x` directly.